pylops/signalprocessing/fourierradon2d.py

```diff
@@ -0,0 +1,296 @@
+            e^{-j \omega_i p_{x,N_p} x^l_1}  & e^{-j \omega_i p_{x,N_p} x^l_2} &  \ldots & e^{-j \omega_i p_{x,N_p} x^l_{N_x}}  \\
+        \end{bmatrix}
+        \begin{bmatrix}
+            d(x_1, \omega_i)  \\
+            d(x_2, \omega_i)  \\
+            \vdots          \\
+            d(x_{N_x}, \omega_i)
+        \end{bmatrix}
+
+    where :math:`l=1,2`. Similarly the forward mode is implemented by applying the
+    transpose and complex conjugate of the above matrix to the model transformed to
+    the Fourier domain.
+
+    Refer to [1]_ for more theoretical and implementation details.
+
+    .. [1] Sacchi, M. "Statistical and Transform Methods for
+        Geophysical Signal Processing", 2007.
+
+    """
+
+    def __init__(
+        self,
+        taxis: NDArray,
+        haxis: NDArray,
+        pxaxis: NDArray,
+        nfft: int,
+        flims: Optional[Tuple[int, int]] = None,
+        kind: str = "linear",
+        engine: str = "numpy",
+        num_threads_per_blocks: Tuple[int, int] = (32, 32),
+        dtype: DTypeLike = "float64",
+        name: str = "R",
+    ) -> None:
+        # engine
+        if engine not in ["numpy", "numba", "cuda"]:
+            raise NotImplementedError("engine must be numpy or numba or cuda")
+        if engine == "numba" and jit_message is not None:
+            engine = "numpy"
+
+        # dimensions and super
+        dims = len(pxaxis), len(taxis)
+        dimsd = len(haxis), len(taxis)
+        super().__init__(dtype=np.dtype(dtype), dims=dims, dimsd=dimsd, name=name)
+
+        # other input params
+        self.taxis, self.haxis = taxis, haxis
+        self.nh, self.nt = self.dimsd
+        self.px = pxaxis
+        self.npx, self.nfft = self.dims[0], nfft
+        self.dt = taxis[1] - taxis[0]
+        self.dh = haxis[1] - haxis[0]
+        self.f = np.fft.rfftfreq(self.nfft, d=self.dt)
+        self.nfft2 = len(self.f)
+        self.cdtype = get_complex_dtype(dtype)
+        self.flims = (0, self.nfft2) if flims is None else flims
+
+        if kind == "parabolic":
+            self.haxis = self.haxis**2
+
+        # create additional input parameters for engine=cuda
+        if engine == "cuda":
+            self.num_threads_per_blocks = num_threads_per_blocks
+            (
+                num_threads_per_blocks_hpx,
+                num_threads_per_blocks_f,
+            ) = num_threads_per_blocks
+            num_blocks_px = (
+                self.dims[0] + num_threads_per_blocks_hpx - 1
+            ) // num_threads_per_blocks_hpx
+            num_blocks_h = (
+                self.dimsd[0] + num_threads_per_blocks_hpx - 1
+            ) // num_threads_per_blocks_hpx
+            num_blocks_f = (
+                self.dims[1] + num_threads_per_blocks_f - 1
+            ) // num_threads_per_blocks_f
+            self.num_blocks_matvec = (num_blocks_h, num_blocks_f)
+            self.num_blocks_rmatvec = (num_blocks_px, num_blocks_f)
```

Comment:
This has the capacity to create more threads than the GPU can support. When dealing with large arrays, a common approach is to choose a fixed number of blocks/threads (ideally maximizing occupancy) and making each thread responsible not for one point in the array, but for as many as it takes to process the whole large array. For now we can leave it like this, but I suggest using this technique in the future. I discuss it here: https://towardsdatascience.com/cuda-by-numba-examples-1-4-e0d06651612f

Reply:
This makes sense, and I think I should spend some time implementing it (also as an exercise for myself ;). 

Right now we ask the user to provide `num_threads_per_blocks`, and we find `num_blocks_matvec` and `num_blocks_rmatvec` to ensure that the grid covers the entire 2d array we want to process... which is easy to do as we know nt/nx/npx. In the case you suggest, do you know of a simple way to check internally how many blocks can be created to not exceed the amount of threads that the GPU can support. Basically, in your tutorial, you do:
```
threads_per_block = 256
blocks_per_grid_gs = 32 * 80  # Use 32 * multiple of streaming multiprocessors
# 32 * 80 * 256 < 1_000_000 so one thread will process more than one array element
```
but surely `32 * multiple of streaming multiprocessors` changes from GPU family to GPU family... do you know any cupy/numba methods to easily get that at runtime? If not, I can dig and find out, but if you know already that would spare me some time ;)

Reply:
Unfortunately, no, cupy and numba dont provide that. It can be obtained via pycuda or cuda-python. I think most ppl still use one-thread-one-index instead of grid stride loops because they don't often encounter massive arrays. In any case, here is a snipper of how the number of blocks would be calculated for maximum occupancy. I don't know how to do that in python though :(

https://github.com/te42kyfo/gpu-benches/blob/master/um-stream/main.cu#L28-L47

Reply:
Alright, let me get the codes with grid stride loops in shape in a separate PR and then we can think of a smart way to handle this, ideally with minimal involvement from the user :)

Reply:
Interesting, for once ChatGPT seemed to have helped. I got the same info you have in your appendix of the Medium blog post using:
```
### Cupy
import cupy as cp

# Get the device (GPU)
device = cp.cuda.Device(0)  # 0 is the device ID, change if needed
device_props = cp.cuda.runtime.getDeviceProperties(0)

# Get name
device_name = device_props['name']

# Get number of MPs
multiprocessor_count = device_props['multiProcessorCount']

# Get the maximum number of threads per block
max_threads_per_block = device.attributes['MaxThreadsPerBlock']

# Get the maximum grid dimensions
max_grid_dim_x = device.attributes['MaxGridDimX']
max_grid_dim_y = device.attributes['MaxGridDimY']
max_grid_dim_z = device.attributes['MaxGridDimZ']

print(f"Device Name: {device_name}")
print(f"Maximum number of multiprocessors: {multiprocessor_count}")
print(f"Maximum threads per block: {max_threads_per_block}")
print(f"Maximum grid dimension (X): {max_grid_dim_x}")
print(f"Maximum grid dimension (Y): {max_grid_dim_y}")
print(f"Maximum grid dimension (Z): {max_grid_dim_z}")
```

```
### Numba
from numba import cuda

# Get the current device (GPU)
device = cuda.get_current_device()

# Get name
device_name = device.name

# Get number of MPs
multiprocessor_count = device.MULTIPROCESSOR_COUNT

# Get the maximum number of threads per block
max_threads_per_block = device.MAX_THREADS_PER_BLOCK

# Get the maximum grid dimensions
max_grid_dim_x = device.MAX_GRID_DIM_X
max_grid_dim_y = device.MAX_GRID_DIM_Y
max_grid_dim_z = device.MAX_GRID_DIM_Z

print(f"Device Name: {device_name}")
print(f"Maximum number of multiprocessors: {multiprocessor_count}")
print(f"Maximum threads per block: {max_threads_per_block}")
print(f"Maximum grid dimension (X): {max_grid_dim_x}")
print(f"Maximum grid dimension (Y): {max_grid_dim_y}")
print(f"Maximum grid dimension (Z): {max_grid_dim_z}")

```